providers/implementations/encode_decode/decode_der2key.c

```diff
@@ -284,60 +272,73 @@ static int der2key_decode(void *vctx, OSSL_CORE_BIO *cin, int selection,
         return 0;
     }
 
-    SET_ERR_MARK();
-    if (!read_der(ctx->provctx, cin, &der, &der_len))
+    ERR_set_mark();
+    ok = read_der(ctx->provctx, cin, &der, &der_len);
+    ERR_pop_to_mark();
```

Comment:
This seems to me overly harsh to silence all read_der() errors. Actually in many cases the read_der() errors are quite good indicator of what is wrong.  Perhaps we should signal fatal errors from read_der() and those should not be cleared? I.e. make return value from read_der a 3 state. In concrete - if I am unable to read anything from cin (at EOF, or some other error), that would be a fatal error.

Reply:
> if I am unable to read anything from cin (at EOF, or some other error), that would be a fatal error.

How do you detect "couldn't read anything" as opposed to "could read but that isn't DER"?

Reply:
I do understand that it's nice if we detect things like corrupt disks, no bytes to read at all and all that, and break the decoding process on those...  but it's not as shallow as checking the EOF or error in `read_der()`, as those are flagged *after* an attempted read.  This means that we have to detect those within the ASN1 library, or go digging through the error queue for certain BIO errors (oh yay, that again...).

Note that from the ASN1 library, at least as far as I know, there's *no* difference that I know of between an error because there were zero bytes to read or an error because there were bytes but none of them could be parsed as a full DER encoded object.

So the question is where to strike a balance between "proper" detection (how deep do you want to go into that rabbit hole) and simply let all decoder implementations go through the same and just pass the problem forward until the process just finds that no final object was constructed...

Reply:
Complete disagree there. There is a huge difference between you gave me nothing and you gave me something I couldn't understand. Right now this approach has lead us to a position where we aren't able to meaningfully report anything useful to the human that can do something about it.

Reply:
This has me think that the whole set_mark / pop_to_mark / clear_last_mark dance could be done in `decoder_process()`, based on the value returned by the currently called decoder implementation (clear_last_mark on false, pop_to_mark on true)